src/java/org/apache/cassandra/index/sai/disk/v1/postings/PostingsWriter.java

```diff
@@ -0,0 +1,316 @@
+        int size = 0;
+        while ((posting = postings.nextPosting()) != PostingList.END_OF_STREAM)
+        {
+            writePosting(posting);
+            size++;
+            totalPostings++;
+        }
+
+        assert size > 0 : "No postings were written";
+
+        finish();
+
+        final long summaryOffset = dataOutput.getFilePointer();
+        writeSummary(size);
+        return summaryOffset;
+    }
+
+    public long getTotalPostings()
+    {
+        return totalPostings;
+    }
+
+    private void writePosting(long posting) throws IOException
+    {
+        if (lastPosting == Long.MIN_VALUE)
+        {
+            firstPosting = posting;
+            deltaBuffer[bufferUpto++] = 0;
+        }
+        else
+        {
+            if (posting < lastPosting)
+                throw new IllegalArgumentException(String.format(POSTINGS_MUST_BE_SORTED_ERROR_MSG, posting, lastPosting));
+            long delta = posting - lastPosting;
+            maxDelta = max(maxDelta, delta);
+            deltaBuffer[bufferUpto++] = delta;
+        }
+        lastPosting = posting;
+
+        if (bufferUpto == blockSize)
+        {
+            addBlockToSkipTable();
+            writePostingsBlock();
+            resetBlockCounters();
+        }
+    }
+
+    private void finish() throws IOException
+    {
+        if (bufferUpto > 0)
+        {
+            addBlockToSkipTable();
+            writePostingsBlock();
+        }
+    }
+
+    private void resetBlockCounters()
+    {
+        firstPosting = Long.MIN_VALUE;
+        bufferUpto = 0;
+        maxDelta = 0;
+    }
+
+    private void addBlockToSkipTable()
+    {
+        blockOffsets.add(dataOutput.getFilePointer());
+        blockMaximumPostings.add(lastPosting);
+    }
+
+    private void writeSummary(int exactSize) throws IOException
+    {
+        dataOutput.writeVInt(blockSize);
+        dataOutput.writeVInt(exactSize);
+        writeSkipTable();
+    }
+
+    private void writeSkipTable() throws IOException
+    {
+        assert blockOffsets.size() == blockMaximumPostings.size();
+        dataOutput.writeVInt(blockOffsets.size());
+
+        // compressing offsets in memory first, to know the exact length (with padding)
+        inMemoryOutput.reset();
+
+        writeSortedFoRBlock(blockOffsets, inMemoryOutput);
+        dataOutput.writeVLong(inMemoryOutput.getFilePointer());
+        inMemoryOutput.writeTo(dataOutput);
+        writeSortedFoRBlock(blockMaximumPostings, dataOutput);
+    }
+
+    private void writePostingsBlock() throws IOException
+    {
+        final int bitsPerValue = maxDelta == 0 ? 0 : DirectWriter.unsignedBitsRequired(maxDelta);
+
+        assert DirectReaders.SUPPORTED_BITS_PER_VALUE.contains(bitsPerValue) :
+        "Unsupported bits per value of " + bitsPerValue + " bits. Supported bits per value are: " +
+        DirectReaders.SUPPORTED_BITS_PER_VALUE.stream().map(i -> Integer.toString(i)).collect(Collectors.joining(", "));
```

Comment:
nit: Should we just pre-calculate this and store it in a constant in `DirectReaders` (or wherever makes the most sense)?

Reply:
I've added it to the `DirectReaders`.